### ISSUE TITLE
AmberParser bug with Ambertools16 gaff.dat

### ISSUE THREAD

**Issue by bas-rustenburg** [choderalab/openmoltools#235]
I'm trying to compile an ffxml file from two gaff mol2 files and frcmod files. It seems to be failing somewhere under the hood in the amber parser, I think the angles get read incorrectly somewhere.

I'm using version 0.7.3, attached is a minimal example. 

[openmoltools-min-example.zip](https://github.com/choderalab/openmoltools/files/569763/openmoltools-min-example.zip)

```
(constph) bas-pycharm@src:~/openmoltools-min-example$ python min-example.py
/home/rustenburg/anaconda/envs/constph/lib/python3.5/site-packages/pandas/io/parsers.py:1728: FutureWarning: split() requires a non-empty pattern match.
  yield pat.split(line.strip())
/home/rustenburg/anaconda/envs/constph/lib/python3.5/site-packages/pandas/io/parsers.py:1730: FutureWarning: split() requires a non-empty pattern match.
  yield pat.split(line.strip())
Traceback (most recent call last):
  File "min-example.py", line 2, in <module>
    omt.utils.create_ffxml_file(["isomer_1.gaff.mol2","isomer_2.gaff.mol2"], ["isomer_1.frcmod", "isomer_2.frcmod"] , ffxml_filename="intermediate.xml")
  File "/home/rustenburg/anaconda/envs/constph/lib/python3.5/site-packages/openmoltools/utils.py", line 211, in create_ffxml_file
    ffxml_stream = parser.generate_xml()
  File "/home/rustenburg/anaconda/envs/constph/lib/python3.5/site-packages/openmoltools/amber_parser.py", line 392, in generate_xml
    theta = float(angle[4]) * math.pi / 180.0
ValueError: could not convert string to float: '95.99calculated_based_on_C#C-SH'
```

**Comment by bas-rustenburg** [comment]
The problem might be occurring in this line: https://github.com/choderalab/openmoltools/blob/master/openmoltools/amber_parser.py#L313

I'm not even sure where this "95.99calculated_based_on_C#C-SH" value comes from. Is it some gaff resource file? I'm using Ambertools16, though previously I was relying on Ambertools15 and I haven't seen this issue. Has something changed about `gaff.dat`?

**Comment by bas-rustenburg** [comment]
From gaff.dat: `c1-sh-hs   48.14       95.99calculated_based_on_C#C-SH         0`

I dont think split will work on this line. Is gaff.dat a fixed-width format? If we know the format specification, maybe it could be use to make the parser more robust in the future. In the meantime I'll try ambertools15 again.

**Comment by bas-rustenburg** [comment]
I don't see the error using `ambermini` 15.0.4, and it occurs when using 16.16.0.

**Comment by davidlmobley** [comment]
@bas-rustenburg - you should cross-compare gaff.dat between the two `ambermini` versions. It looks to me like someone introduced a format error or typo between versions. You'll also have to consult the AMBER docs as to the spec for these files. I believe this must be fixed width. But the specs come up easily via Google.

**Comment by jchodera** [comment]
Tagging @lnaden who did the `ambermini` update.

**Comment by Lnaden** [comment]
Yeah, I see the file. It looks like the comment on that 1 line got shifted so there is no longer a space between the last data point and the comment.

**Comment by bas-rustenburg** [comment]
Yeah, these lines are the culprit in my case:

Ambermini 15.0.4:
`c1-sh-hs   48.23       95.99    calculated_based_on_C#C-SH       0`
Ambermini 16.16.0
`c1-sh-hs   48.14       95.99calculated_based_on_C#C-SH         0` 


Format specified here. Not sure what language the format is in, is that fortran? 

http://ambermd.org/formats.html#parm.dat

**Comment by Lnaden** [comment]
Its FORTRAN string formatting syntax, yes. And that bug/line was introduced as part of AmberTools16, not something unique to AmberMini

**Comment by jchodera** [comment]
Confirmed: This is actually the line in `AmberTools`. 

Given that this line conforms to the specification, the bug is with our parser, not with `ambermini`.

We should update the parser to match the formatting. @rafwiewiora may have already updated the AMBER conversion scripts---we might be able to borrow the updates from there.

**Comment by Lnaden** [comment]
For now I think it would be faster to just pin AmberTools 15.0.4 and put a to-do on changing the parser to fixed width

**Comment by davidlmobley** [comment]
> For now I think it would be faster to just pin AmberTools 15.0.4 and put a to-do on changing the parser to fixed width

Please don't, at least not in `openmoltools`. We're doing some testing with GAFF2 on this end and it's only available in AmberTools 16+. If you want to pin to 15.0.4 somewhere else that doesn't affect us, fine, but you'll break stuff we're doing if you pin openmoltools to that.

**Comment by Lnaden** [comment]
Noted, we won't pin to the older version. We're talking internally how to fix this @rafwiewiora may have a solution as @jchodera pointed out.

**Comment by rafwiewiora** [comment]
Ok, so I've converted the AmberTools16 `gaff.dat` recently with my conversion script, but I'm going to have to double check whether there really was no problem given that lack of a space, i.e. whether `ParmEd` reads that ok. 

And will merge it into https://github.com/choderalab/openmm-forcefields (currently sitting here https://github.com/choderalab/openmm/tree/forcefield_conversion_bonanza_amber16/devtools/forcefield-scripts/amber) later today for you guys to use. 

Also will just point out that I realized while using this new `gaff.dat` that it was not the same as the `gaff.dat` in AmberTools15 in terms of some parameters - I think it was some bond and angle parameters that were updated.
